### PR TITLE
Adding searching capabilities

### PR DESCRIPTION
**Pull request by raul-gracia**
## Description
This adds the ability to search by URN, Email, Passport and Full Name

## Trello Card Link

https://trello.com/c/PeGkGRe2

### CODE REVIEW

**Comment by raul-gracia** [comment]
[Applications List: Search Capabilities](https://trello.com/c/PeGkGRe2/123-applications-list-search-capabilities)

**Comment by raul-gracia** [comment]
[Applications List: Search Capabilities](https://trello.com/c/PeGkGRe2/123-applications-list-search-capabilities)

**Review by padv2010 — DISMISSED** [review]
I am finding a few foundational issues:

1. tests are incomplete and do not cover all searching capabilities
2. tests are done at the system level, which slow down performance and makes testing and maintenance difficult.
3. we are using a scope to encapsulate a complex search, which is polluting a class with responsibilities that probably belong in another place.

**Comment by raul-gracia** [comment]
1. As far as I know the test covers the search functionality by URN, Email, passport number and individual given name and family name. Would you like to specify which capability of the search has not been covered?
2. As far as I know, and per the feedback in other PRs, systems tests have always been required. I can add extra tests to cover the scope at the unit test level if it is necessary.
3. IMHO a simple search by 4 individual fields where 3 of them reside on the same model doesn't qualify as a complex query, and is easy to read and understand what is happening. Using a scope allows us to have a straightforward code on the controller and is very simple to use the scope to obtain the results.

**Comment by padv2010** [comment]
> As far as I know the test covers the search functionality by URN, Email, passport number and individual given name and family name. Would you like to specify which capability of the search has not been covered?

Sure, I have found these missing tests, although might be more: 

1. Search tests do not include ignoring capital letters on `:term`. Tests will pass if you just remove `LOWER` and `downcase` from the query. This seems important and is not tested.
2. Search return incomplete elements, but this is not tested. For example, if there is an applicant with your name, and you enter `Ra`, your name will appear. This is part of the query that you have implemented but there is no test about it.
3. Search return full names (given name and family name concatenated). If you search by Pablo Manrubia, my name would appear. If you remove this line from the query: `LOWER(CONCAT(applicants.given_name, ' ', applicants.family_name)) LIKE :term OR` Test will still pass, and it would not be expected. It should have failed.
4. On top of this, your system tests are passing only because you added 20 applicants and pagination is playing a role. Your assertions are only checking that the application exist, but not testing the filtering, so you can have false positives. This means, for example, that if you decide to change the order of the query, the query would not be tested at all. Which is not great. 

Overall, it seems to me that it has not been Unit tested, nor applied TDD in its development and there are gaps (missing tests and lack of documentation via tests)

> As far as I know, and per the feedback in other PRs, systems tests have always been required. I can add extra tests to cover the scope at the unit test level if it is necessary.

You need both; you can read about different testing levels by googling `testing pyramid`

> IMHO a simple search by 4 individual fields where 3 of them reside on the same model doesn't qualify as a complex query, and is easy to read and understand what is happening. Using a scope allows us to have a straightforward code on the controller and is very simple to use the scope to obtain the results.

If you add unit tests, which will also document the requirements for testing, you will find that you are polluting the main class with unnecessary information; we are kicking off the journey of the God Object anti pattern. The `Application` knows too much about searching and it seems that the Single Responsability Principle is also being broken.

Even using a scope, which it is fine here, there are other ways to hide the implementation details. You don't only need pass block with complex SQL embedded. Rails allows you other approaches.

For example, this one:

```ruby
scope :search, Applications::Search
```

This will immediately:

1. Let you do proper TDD and Unit testing at the scope level (proper isolation)
2. Let the `Application` class worry about the `ability` to search, but not on the details
3. Remove the clutter of inner SQL in a scope in the most important class of the domain: the `Application`
4. DRY-up the system tests which are slow and not intended (by design) for covering all cases.

As another minor detail, and for your information, or for future readers, the query that you have included, will always perform a full scan on the database because you are modifying the columns with the `LOWER` keyword. Not a big deal, but something worth having in mind for the future in case performance degrades throughout the years.

If hope this clarifies my feedback Raul, but feel free to reach out in case you want more details. 

As there seem to be some urgency into merging this, I would suggest reaching out to @duncanjbrown in case this has to be merged.

File: app/models/application.rb [hunk 1]
```diff
@@ -24,6 +24,21 @@ class Application < ApplicationRecord
 
   scope :submitted, -> { where.not(urn: nil) }
 
+  scope :search, lambda { |term|
+                   return if term.blank?
+
+                   term = "%#{term.downcase}%"
+                   joins(:applicant).where(
+                     "LOWER(applications.urn) LIKE :term OR
```
Comment (duncanjbrown) [inline app/models/application.rb:32]: I think possible to avoid these `LOWER`s simply by using `ILIKE`?

File: spec/features/admin_console/applications_list_spec.rb [hunk 2]
```diff
@@ -19,8 +19,27 @@
     then_i_can_see_the_pagination
   end
 
+  it "allows searching" do
```
Comment (duncanjbrown) [inline spec/features/admin_console/applications_list_spec.rb:22]: We're making multiple e2e test requests to effectively unit test a single scope inside Application. There's definitely value in a system spec here but I think better to test the edge cases in a unit. Kicking most of these assertions out of here and moving them into a unit test on Application would be both faster and cleaner? Could include case insensitivity in there too.

File: app/models/application.rb [hunk 3]
```diff
@@ -24,6 +24,21 @@ class Application < ApplicationRecord
...
+       LOWER(applicants.email_address) LIKE :term OR
+       LOWER(applicants.passport_number) LIKE :term OR
+       LOWER(CONCAT(applicants.given_name, ' ', applicants.family_name)) LIKE :term OR
```
[the remark this answers is not in the document]
Comment (raul-gracia) [reply app/models/application.rb:35]: This covers the case when they search for the full name with the space in between, I tested it and it won't be covered by searching only for given name or family name.

File: spec/features/admin_console/applications_list_spec.rb (hunk 2 above)
Comment (raul-gracia) [reply spec/features/admin_console/applications_list_spec.rb:22]: I've left only one of the tests here and I've moved the rest of the tests including case insensitivity to the unit test level.

**Review by duncanjbrown — APPROVED** [review]
Thank you!
